### PR TITLE
pass `--lib` to `x doc`

### LINKED ISSUE #112082: the docs of rustdoc get overwritten by the bootstrap docs
Since #111976, `bootstrap` itself gets documented. Unfortunately, the package contains a binary crate called `rustdoc` (`src/bootstrap/bin/rustdoc.rs`, *Shim which is passed to Cargo as “rustdoc” when running the bootstrap.*) which clashes with the actual `rustdoc` package (`src/librustdoc`) overwriting the latter since it gets documented last.

See for yourself: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/index.html (this ain't *rustdoc*).

CC @jyn514 (#111976).

@rustbot label C-bug T-bootstrap A-docs A-contributor-roadblock

### PR DESCRIPTION
**Pull request by eval-exec**
This PR want to close #112082

### CODE REVIEW

**Review by onur-ozkan — CHANGES_REQUESTED** [review]
Thank you for the fix.

Can you update the following as well?
https://github.com/rust-lang/rust/blob/0939ec13d88dfafcbb7f25314bd0d2f1519bf0d5/src/bootstrap/builder.rs#L1576

**Comment by jyn514** [comment]
I'm short on time for reviews, please don't assign me.

**Comment by jyn514** [comment]
@eval-exec try `x doc --stage 0 bootstrap rustdoc`

**Comment by ozkanonur** [comment]
> r? @jyn514 @ozkanonur
> 
> @jyn514 @ozkanonur Request for your review heart_decoration
> 
> @rustbot review

jfyi: In most cases, we do not reassign the reviewer unless there is a specific reason to do so :)

File: src/bootstrap/Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ test = false
 
 [[bin]]
 name = "rustdoc"
```
Comment (ozkanonur) [inline src/bootstrap/Cargo.toml:23]: this needs to be `rustdoc-shim` in order to fix https://github.com/rust-lang/rust/issues/112082

Comment (ozkanonur) [reply src/bootstrap/Cargo.toml:23]: which requires updating https://github.com/rust-lang/rust/blob/a9baa16482ba4100df529ccba39c787f27ad0475/src/bootstrap/dist.rs#L2312-L2314 too

**Comment by ozkanonur** [comment]
If things goes more complicated than we expected, we can implement the first solution which was `pass --lib to cargo so it doesn't try to document the binaries`

**Comment by eval-exec** [comment]
> If things goes more complicated than we expected, we can implement the first solution which was `pass --lib to cargo so it doesn't try to document the binaries`

It's hard, `rustdoc`'s doc is still as same as `rustdoc-shim`
image
image

I'd like to pass `--lib` to cargo. :smile:

**Comment by ozkanonur** [comment]
> It's hard, rustdoc's doc is still as same as rustdoc-shim

It's probably because you didn't build `rustdoc` again after updating binary name. So this was from the previous build. Regenerate the docs after `x clean`, then it should do the job.

**Comment by eval-exec** [comment]
`./x doc` is slow on my machine, I'm going to sleep and continue this PR 7 hours later.

**Comment by ozkanonur** [comment]
> ./x doc is slow on my machine, I'm going to sleep and continue this PR 7 hours later.

No worries, I tested it and it works well :)

@bors r+ rollup

File: src/bootstrap/doc.rs [hunk 2]
```diff
@@ -834,6 +834,7 @@ macro_rules! tool_doc {
                 cargo.arg("-Zskip-rustdoc-fingerprint");
                 // Only include compiler crates, no dependencies of those, such as `libc`.
                 cargo.arg("--no-deps");
+                cargo.arg("--lib");
```
Comment (jyn514) [inline src/bootstrap/doc.rs:837]: this is passed for all tools, not just bootstrap. was that intentional?

Comment (ozkanonur) [reply src/bootstrap/doc.rs:837]: oops, missed it. Will fix it

Comment (eval-exec) [reply src/bootstrap/doc.rs:837]: I'm sorry, I made a mistake. Thank you for your correction.

**Pull request merged**